### PR TITLE
CLI update from diff to compare & add TableComparator

### PR DESCRIPTION
**Pull request by JihwanEom**
### Summary
- CLI update from diff to compare
- Add TableComparator functionality as new method
- Refactor existing features for HLOps structure 
- Fix broken links in project.rst

### How to test
```
datum compare tests/assets/cityscapes_dataset/train_dataset/ tests/assets/ade20k2020_dataset/

# Save output json and testable
datum compare tests/assets/cityscapes_dataset/train_dataset/ tests/assets/ade20k2020_dataset/ --output-dir compare_test
```

### Output preview
```
$ datum compare tests/assets/cityscapes_dataset/train_dataset/ tests/assets/ade20k2020_dataset/
High-level comparison:
+--------------------------+-------------------------------------+------------------------+
| Field                    | First                               | Second                 |
+==========================+=====================================+========================+
| Format                   | cityscapes                          | ade20k2020             |
+--------------------------+-------------------------------------+------------------------+
| Number of classes        | 20                                  | 4                      |
+--------------------------+-------------------------------------+------------------------+
| Common classes           | car, person                         | car, person            |
+--------------------------+-------------------------------------+------------------------+
| Classes                  | background, bicycle, building, bus, | car, door, person, rim |
|                          | car, fence, motorcycle, person,     |                        |
|                          | pole, rider, road, sidewalk, sky,   |                        |
|                          | terrain, trafficlight, trafficsign, |                        |
|                          | train, truck, vegetation, wall      |                        |
+--------------------------+-------------------------------------+------------------------+
| Images count             | 4                                   | 3                      |
+--------------------------+-------------------------------------+------------------------+
| Unique images count      | 1                                   | 1                      |
+--------------------------+-------------------------------------+------------------------+
| Repeated images count    | 1                                   | 1                      |
+--------------------------+-------------------------------------+------------------------+
| Annotations count        | 8                                   | 22                     |
+--------------------------+-------------------------------------+------------------------+
| Unannotated images count | 0                                   | 0                      |
+--------------------------+-------------------------------------+------------------------+

Mid-level comparison:
+-------------------------------------+--------------------------+--------------------------+
| Field                               | First                    | Second                   |
+=====================================+==========================+==========================+
| test - Image Mean                   | 1.00,   1.00,   1.00     |                          |
+-------------------------------------+--------------------------+--------------------------+
| test - Image Std                    | 0.00,   0.00,   0.00     |                          |
+-------------------------------------+--------------------------+--------------------------+
| train - Image Mean                  | 1.00,   1.00,   1.00     |                          |
+-------------------------------------+--------------------------+--------------------------+
| train - Image Std                   | 0.00,   0.00,   0.00     |                          |
+-------------------------------------+--------------------------+--------------------------+
| val - Image Mean                    | 1.00,   1.00,   1.00     |                          |
+-------------------------------------+--------------------------+--------------------------+
| val - Image Std                     | 0.00,   0.00,   0.00     |                          |
+-------------------------------------+--------------------------+--------------------------+
| dataset - Image Mean                |                          | 1.00,   1.00,   1.00     |
+-------------------------------------+--------------------------+--------------------------+
| dataset - Image Std                 |                          | 0.00,   0.00,   0.00     |
+-------------------------------------+--------------------------+--------------------------+
| dataset_with_meta_file - Image Mean |                          | 1.00,   1.00,   1.00     |
+-------------------------------------+--------------------------+--------------------------+
| dataset_with_meta_file - Image Std  |                          | 0.00,   0.00,   0.00     |
+-------------------------------------+--------------------------+--------------------------+
| Label - background                  | imgs: 4, percent: 0.5000 |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - bicycle                     |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - building                    |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - bus                         |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - car                         |                          | imgs: 6, percent: 0.2727 |
+-------------------------------------+--------------------------+--------------------------+
| Label - fence                       |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - motorcycle                  |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - person                      | imgs: 1, percent: 0.1250 | imgs: 8, percent: 0.3636 |
+-------------------------------------+--------------------------+--------------------------+
| Label - pole                        |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - rider                       |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - road                        |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - sidewalk                    |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - sky                         |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - terrain                     |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - trafficlight                |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - trafficsign                 |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - train                       | imgs: 1, percent: 0.1250 |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - truck                       | imgs: 1, percent: 0.1250 |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - vegetation                  |                          |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - wall                        | imgs: 1, percent: 0.1250 |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - door                        |                          | imgs: 6, percent: 0.2727 |
+-------------------------------------+--------------------------+--------------------------+
| Label - rim                         |                          | imgs: 2, percent: 0.0909 |
+-------------------------------------+--------------------------+--------------------------+

Low-level comparison:
+-----------------+---------+
| Field           |   Value |
+=================+=========+
| Covariate shift |   0     |
+-----------------+---------+
| Label shift     |   0.999 |
+-----------------+---------+

```

### Test coverage
```
Name                                                                                   Stmts   Miss Branch BrPart  Cover
------------------------------------------------------------------------------------------------------------------------
datumaro/plugins/comparator.py                                                           321     37    120      8    87%
```

### Checklist

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [x] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [x] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: datumaro/cli/commands/compare.py [hunk 1]
```diff
@@ -221,44 +219,18 @@ def diff_command(args):
     if args.method is ComparisonMethod.equality:
         if args.ignore_field:
             args.ignore_field = eq_default_if
-        comparator = ExactComparator(
```
Comment (vinnamkim) [inline datumaro/cli/commands/compare.py]: Why do you remove the existing features?

**Review by vinnamkim — CHANGES_REQUESTED** [review]

File: datumaro/cli/commands/compare.py (hunk 1 above)
Comment (JihwanEom) [reply datumaro/cli/commands/compare.py]: Thank you for comment, I'll revert existed features and combine with new feature.

**Comment by JihwanEom** [comment]
PR Test / pr test (macOS-11, Python3.8) is failing due to below issue:
```
2023-05-26T01:14:09.2846690Z ========== 1149 passed, 19 skipped, 94 warnings in 386.18s (0:06:26) ===========
2023-05-26T01:14:13.8925520Z /Users/runner/work/_temp/8441724d-c583-4fd5-93bd-665650ace024.sh: line 1:  6254 Segmentation fault: 11  python -m pytest -v tests/unit/ --cov
```
It occurred regardless of new test cases for TableComparator. It's hard to debug limited traceback by Segmentation fault, I'll rebase and check again after merging #1030. Please review code architecture first.

**Comment by JihwanEom** [comment]
> PR Test / pr test (macOS-11, Python3.8) is failing due to below issue:
> 
> ```
> 2023-05-26T01:14:09.2846690Z ========== 1149 passed, 19 skipped, 94 warnings in 386.18s (0:06:26) ===========
> 2023-05-26T01:14:13.8925520Z /Users/runner/work/_temp/8441724d-c583-4fd5-93bd-665650ace024.sh: line 1:  6254 Segmentation fault: 11  python -m pytest -v tests/unit/ --cov
> ```
> 
> It occurred regardless of new test cases for TableComparator. It's hard to debug limited traceback by Segmentation fault, I'll rebase and check again after merging #1030. Please review code architecture first.

Resolved by 3113cba73eeb204095bbf9fe35d5c3240f3be6bf

**Review by wonjuleee — COMMENTED** [review]
Could you clarify the differences between `DistanceComparator`, `EqualityComparator`, and `TableComparator`?

**Comment by JihwanEom** [comment]
> Could you clarify the differences between `DistanceComparator`, `EqualityComparator`, and `TableComparator`?

The `DistanceComparator` supports item-wise comparison for the `label`, `bbox`, `polygon`, and `mask` annotation types, and it saves a confusion matrix for each type.
The `EqualityComparator` is used to compare whether the annotations in two given datasets are exactly the same.
The `TableComparator` is a new feature proposed in this PR. It provides a high-level view of the datasets, not on an item-wise basis, but on the overall perspective. `TableComparator` represents the dataset's format, number of classes, common classes, and image count. In the mid-level view, it presents image and annotation statistics for each subset. In the low-level comparison, it utilizes a shift analyzer to represent covariate shift and label shift.

Below are examples of the output for each comparator.
```
datum compare tests/assets/ade20k2017_dataset/ tests/assets/ade20k2020_dataset/ --method distance --output-dir output/distance
datum compare tests/assets/ade20k2017_dataset/ tests/assets/ade20k2020_dataset/ --method equality --output-dir output/equality 
datum compare tests/assets/ade20k2017_dataset/ tests/assets/ade20k2020_dataset/ --method table --output-dir output/table
```

image

### DistanceComparator
`mask_confusion.png`
image
`polygon_confusion.png`
image

### EqualityComparator
`equality_compare.json`
```
{
  "mismatches": [],
  "a_extra_items": [],
  "b_extra_items": [],
  "errors": [
    {
      "type": "labels",
      "message": "Lists differ: [Labe[22 chars]ame='sky', parent='', attributes=set()), Label[203 chars]t())] != [Labe[22 chars]ame='car', parent='', attributes=set()), Label[194 chars]t())]\n\nFirst differing element 0:\nLabelCategories.Category(name='sky', parent='', attributes=set())\nLabelCategories.Category(name='car', parent='', attributes=set())\n\n- [LabelCategories.Category(name='sky', parent='', attributes=set()),\n?                                 ^^^\n\n+ [LabelCategories.Category(name='car', parent='', attributes=set()),\n?                                 ^^^\n\n   LabelCategories.Category(name='person', parent='', attributes=set()),\n-  LabelCategories.Category(name='license plate', parent='', attributes=set()),\n?                                 ^^^^^^^^^^^^^\n\n+  LabelCategories.Category(name='door', parent='', attributes=set()),\n?                                 ^^^^\n\n   LabelCategories.Category(name='rim', parent='', attributes=set())]"
    }
  ]
}
```

### TableComparator
`table_compare.txt`
```
High-level Comparison:
+--------------------------+---------------------------------+------------------------+
| Field                    | First                           | Second                 |
+==========================+=================================+========================+
| Format                   | ade20k2017                      | ade20k2020             |
+--------------------------+---------------------------------+------------------------+
| Number of classes        | 4                               | 4                      |
+--------------------------+---------------------------------+------------------------+
| Common classes           | person, rim                     | person, rim            |
+--------------------------+---------------------------------+------------------------+
| Classes                  | license plate, person, rim, sky | car, door, person, rim |
+--------------------------+---------------------------------+------------------------+
| Images count             | 3                               | 3                      |
+--------------------------+---------------------------------+------------------------+
| Unique images count      | 1                               | 1                      |
+--------------------------+---------------------------------+------------------------+
| Repeated images count    | 1                               | 1                      |
+--------------------------+---------------------------------+------------------------+
| Annotations count        | 10                              | 22                     |
+--------------------------+---------------------------------+------------------------+
| Unannotated images count | 0                               | 0                      |
+--------------------------+---------------------------------+------------------------+

Mid-level Comparison:
+-------------------------------------+--------------------------+--------------------------+
| Field                               | First                    | Second                   |
+=====================================+==========================+==========================+
| dataset - Image Mean                | 1.00,   1.00,   1.00     | 1.00,   1.00,   1.00     |
+-------------------------------------+--------------------------+--------------------------+
| dataset - Image Std                 | 0.00,   0.00,   0.00     | 0.00,   0.00,   0.00     |
+-------------------------------------+--------------------------+--------------------------+
| dataset_with_meta_file - Image Mean | 1.00,   1.00,   1.00     | 1.00,   1.00,   1.00     |
+-------------------------------------+--------------------------+--------------------------+
| dataset_with_meta_file - Image Std  | 0.00,   0.00,   0.00     | 0.00,   0.00,   0.00     |
+-------------------------------------+--------------------------+--------------------------+
| Label - license plate               | imgs: 3, percent: 0.3000 |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - person                      | imgs: 3, percent: 0.3000 | imgs: 8, percent: 0.3636 |
+-------------------------------------+--------------------------+--------------------------+
| Label - rim                         | imgs: 1, percent: 0.1000 | imgs: 2, percent: 0.0909 |
+-------------------------------------+--------------------------+--------------------------+
| Label - sky                         | imgs: 3, percent: 0.3000 |                          |
+-------------------------------------+--------------------------+--------------------------+
| Label - car                         |                          | imgs: 6, percent: 0.2727 |
+-------------------------------------+--------------------------+--------------------------+
| Label - door                        |                          | imgs: 6, percent: 0.2727 |
+-------------------------------------+--------------------------+--------------------------+

Low-level Comparison:
+-----------------+---------+
| Field           |   Value |
+=================+=========+
| Covariate shift |    0    |
+-----------------+---------+
| Label shift     |    0.75 |
+-----------------+---------+

```

**Review by wonjuleee — DISMISSED** [review]

**Review by sooahleex — DISMISSED** [review]

File: datumaro/plugins/comparator.py
Comment (vinnamkim) [inline datumaro/plugins/comparator.py:1]: Those classes are not fitted to the definition of Datumaro plugins (loaded by `Environment`). Please move this to somewhere under `datumaro/components`.

File: datumaro/plugins/shift_analyzer.py
Comment (vinnamkim) [inline datumaro/plugins/shift_analyzer.py:1]: Here too

File: datumaro/plugins/comparator.py
Comment (JihwanEom) [reply datumaro/plugins/comparator.py:1]: Fixed in d9f71a5be8f463207bc4fa6221b88e22f91126ed

File: datumaro/plugins/shift_analyzer.py
Comment (JihwanEom) [reply datumaro/plugins/shift_analyzer.py:1]: Fixed in d9f71a5be8f463207bc4fa6221b88e22f91126ed

File: src/datumaro/plugins/openvino_plugin/shift_launcher.py
Comment (vinnamkim) [inline src/datumaro/plugins/openvino_plugin/shift_launcher.py:1]: It seems that you have two duplicated source file:
1) src/datumaro/plugins/openvino_plugin/shift_launcher.py
2) src/datumaro/plugins/shift_analyzer.py

Comment (JihwanEom) [reply src/datumaro/plugins/openvino_plugin/shift_launcher.py:1]: Removed later one in e9d7a31f102dd8cacdabbf06bd75e0fbd107fed4, thank you!

**Review by vinnamkim — APPROVED** [review]
LGTM.

**Review by sooahleex — APPROVED** [review]
